### PR TITLE
makefiles/utils/variables: add functions to help managing variables

### PR DESCRIPTION
**Pull request by cladmi**
### Contribution description

Add functions to set variables and environment for targets
These functions should help replacing immediate evaluation and global 'export'

#### `target-export-variables`

`target-export-variables <target> <variables>` function that exports a list of variables for some targets

This should help exporting variables only for the required targets so not evaluate `shell` calls for nothing.
Like `make clean` does not need `CFLAGS` but if the variable is exported, it will evaluate which options are supported in `CFLAGS` anyway.

#### `memoized`

`VARIABLE = $(call memoized,VARIABLE,<value>)` function that allows only evaluating `value` when the variable is used like a normal deferred variable, but further evaluations will use an immediate value.

This should help not assigning `shell` commands results with `:=` to call it once only, as it result to evaluating even if not needed. Replacing by a direct `=` would result in the command be executed on each evaluation which is not wanted.
This gives the best of both worlds.

The implementation requires to repeat the variable name in the `call memoized` arguments.

### Testing procedure

Review and run the test in `tests/build_system_utils`.

```
make -C tests/build_system_utils/
```
If it executes without error, it is working.

You can also run it with `QUIET=0` to see the executed commands.

### Issues/PRs references

Tracking: remove harmful use of `export` in make and immediate evaluation #10850

### CODE REVIEW

File: makefiles/utils/variables.mk [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Utilities to set variables and environment for targets
+# These functions should help replacing immediate evaluation and global 'export'
+
+
+# Evaluate a deferred variable only on once on its first usage
```
Comment (smlng) [inline makefiles/utils/variables.mk]: typo: double `on`, remove before `once`

File: makefiles/utils/variables.mk [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# Uses after will be as if it was an immediate evaluation
+# This can replace using `:=` by default
+#
+# The goal is to use it for `shell` commands
+#
+# variable = $(call memoized,<variable>,<value>)
+#
+# Parameters
+#  variable: name on the variable you set
```
Comment (smlng) [inline makefiles/utils/variables.mk]: typo: /on/of/

File: makefiles/utils/variables.mk [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# Utilities to set variables and environment for targets
+# These functions should help replacing immediate evaluation and global 'export'
+
+
+# Evaluate a deferred variable only on once on its first usage
+# Uses after will be as if it was an immediate evaluation
```
Comment (smlng) [inline makefiles/utils/variables.mk]: wording: better to read: `# Uses after [that] will be as ...`

**Comment by cladmi** [comment]
Damn it does not support `date +%s%N` I will see what is available and does not require 1s sleep.

Memoized comes from the concept https://en.wikipedia.org/wiki/Memoization but I welcome alternative names.

**Comment by aabadie** [comment]
> Memoized comes from the concept https://en.wikipedia.org/wiki/Memoization but I welcome alternative names.

I think this name suits perfectly, no need for another one.

File: makefiles/utils/variables.mk (hunk 1 above)
Comment (cladmi) [reply makefiles/utils/variables.mk]: I can see that I piggybacked that change from a wip branch a bit too fast :D

**Comment by cladmi** [comment]
> bash: line 0: test: 1560237203N: integer expression expected
ERROR: 1560237203N >= 1560237203N
make[1]: *** [test-exported-variables] Error 1
Command '/Library/Developer/CommandLineTools/usr/bin/make -C /Volumes/devel/github/smlng/RIOT/makefiles/utils -f test-variables.mk test-exported-variables' failed
bash: line 0: test: 1560237203N: integer expression expected

The error is not in docker but in `osx`.
I could replace it with `date; sleep 1` in `osx` but will see if I find something better.

**Comment by cladmi** [comment]
@smlng I updated to use `python` to give the timestamp in nanoseconds and it worked on `OSx 17.7.0`. Does it work for you too ?

**Comment by smlng** [comment]
yep, python works on macOS  ... alternative would be to use `gdate` on macOS (and other non-Linux systems), but that's not available by default.

**Comment by cladmi** [comment]
> yep, python works on macOS ... alternative would be to use `gdate` on macOS (and other non-Linux systems), but that's not available by default.

Here it is just a test, not the all applications workflow, so I did not look for performance optimization. And it gives a solution without asking for `uname` and if `gdate` is there or not.

**Comment by cladmi** [comment]
@smlng do you agree with the updated changes ?

**Comment by smlng** [comment]
yep, pleas squash

**Comment by smlng** [comment]
please 😄

**Comment by cladmi** [comment]
Squashed.

**Comment by cladmi** [comment]
Thank you for the review. We need to start using it to remove more exports.

**Pull request merged**